### PR TITLE
Add clustercheck configprovider and matching API

### PR DESCRIPTION
**Pull request by xvello**
### What does this PR do?

This PR introduces the communication logic between the node-agents and the cluster-agent, and re-structures the clustercheck classes to best support it. For now, the cluster-agent dispatches all the configs to the node it is running on. Real multi-node dispatching will come in the next PR.

#### Node-agent
Add the `clusterchecks` configprovider (disabled by default) and supporting methods in `pkg/util/clusteragent`.

As discussed in the RFC, it has a `grace_time_seconds` option (default 60 seconds), during which it will keep its current state if it cannot reach a DCA. After this, it will drop its configs.

#### Cluster-agent

- add the `/api/v1/clusterchecks/status/{nodeName}` and `/api/v1/clusterchecks/configs/{nodeName}` endpoints
- add a stub `redirectToLeader` logic, that will be used when leader-election will be implemented, no-op for now
- Rework class scopes, see [updated architecture schema](https://github.com/DataDog/datadog-agent/blob/xvello/ccheck-provider/pkg/clusteragent/clusterchecks/README.md). Keeping dispatching transactions atomic with an internal lock on the stores was too painful. Moving the `clusterStore` as an internal of `dispatcher`, that holds the mutex. `dispatcher` provides accessor methods for `Handler` now.

### CODE REVIEW

File: pkg/clusteragent/clusterchecks/stores.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// clusterStore holds the state of cluster-check management.
+// Lock is to be held by the dispatcher so it can make atomic
+// operations involving several calls.
+type clusterStore struct {
+	sync.RWMutex
+	digestToConfig map[string]integration.Config // All configurations to dispatch
+	digestToNode   map[string]string             // Node running a config
```
Comment (xvello) [inline pkg/clusteragent/clusterchecks/stores.go]: This map alone can allow to rebuild the dispatching state (assuming our digests are stable).
We should be able to store it in a file and restore from it

**Review by hkaj — DISMISSED** [review]

File: pkg/clusteragent/clusterchecks/dispatcher_main.go [hunk 2]
```diff
@@ -41,18 +43,16 @@ func (d *dispatcher) Unschedule(configs []integration.Config) {
...
 // add stores and delegates a given configuration
 func (d *dispatcher) add(config integration.Config) {
 	if !config.ClusterCheck {
 		return // Ignore non cluster-check configs
 	}
 	log.Debugf("dispatching configuration %s:%s", config.Name, config.Digest())
-	d.store.addConfig(config)
+
+	// TODO: add dispatching logic
+	hostname, _ := util.GetHostname()
```
Comment (mfpierre) [inline pkg/clusteragent/clusterchecks/dispatcher_main.go]: If I get this right, currently all cluster checks are dispatched to the same host as the cluster-agent leader?

**Review by mfpierre — COMMENTED** [review]
Nice work 🙌 

Left some comment/questions, I'll probably need a second pass 🔍

File: pkg/clusteragent/clusterchecks/dispatcher_main.go (hunk 2 above)
Comment (xvello) [reply pkg/clusteragent/clusterchecks/dispatcher_main.go]: Yes, this PR only brings in the piping with a placeholder logic, so the next PR can focus on the dispatching logic and be easier to review

File: cmd/cluster-agent/api/v1/clusterchecks.go [hunk 3]
```diff
@@ -14,43 +14,123 @@ import (
...
+	w.WriteHeader(http.StatusNotFound)
+}
+
+// redirectToLeader handles the logic for followers to transparently
+// redirect the requests to the leader cluster-agent
+func redirectToLeader(w http.ResponseWriter, r *http.Request, h *clusterchecks.Handler) bool {
+	if leader := h.ShouldRedirect(); leader != "" {
+		url := r.URL
+		url.Host = leader
+		http.Redirect(w, r, url.String(), http.StatusFound)
```
[the remark this answers is not in the document]
Comment (xvello) [reply cmd/cluster-agent/api/v1/clusterchecks.go]: We're sending a `302` code, the node-agent is responsible to query the leader directly

File: pkg/clusteragent/clusterchecks/stores.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+// getNodeStore retrieves the store struct for a given node name. If the node
+// is not yet registered in the store, an entry will be inserted and returned,
+// or an empty pointer will be returned
+func (s *clusterStore) getNodeStore(nodeName string, create bool) (*nodeStore, bool) {
+	node, ok := s.nodes[nodeName]
+	if ok {
+		return node, true
+	}
+
+	if !create {
```
[the remark this answers is not in the document]
Comment (xvello) [reply pkg/clusteragent/clusterchecks/stores.go]: indeed, that boolean was not very expressive. Introduced `getOrCreateNodeStore` for when we want to register a node if not present

**Review by hkaj — APPROVED** [review]

**Pull request merged**